### PR TITLE
Fix golint failures: pkg/controller/certificates #68026

### PR DESCRIPTION
**Pull request by zerodayz**
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
This PR fix golint issues for pkg/controller/certificates:

Errors from golint:
pkg/controller/certificates/certificate_controller.go:43:6: exported type CertificateController should have comment or be unexported
pkg/controller/certificates/certificate_controller.go:57:1: exported function NewCertificateController should have comment or be unexported
pkg/controller/certificates/certificate_controller.go:212:52: exported func IgnorableError returns unexported type certificates.ignorableError, which can be annoying to use
pkg/controller/certificates/certificate_controller_utils.go:28:1: exported function GetCertApprovalCondition should have comment or be unexported

Which issue(s) this PR fixes:

Signed-off-by: Robin Cernin <cerninr@gmail.com>
#68026

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/controller/certificates/certificate_controller.go
Comment (soltysh) [inline pkg/controller/certificates/certificate_controller.go:219]: I'd prefer we don't expose types we don't want to be exposed just because golint complains. Once made public it's part of the API we'll need to support. So I'd prefer not doing this.

**Review by soltysh — CHANGES_REQUESTED** [review]
/hold

File: pkg/controller/certificates/certificate_controller.go
Comment (awly) [inline pkg/controller/certificates/certificate_controller.go:214]: This should return a regular `error` instead

File: pkg/controller/certificates/certificate_controller.go
Comment (awly) [inline pkg/controller/certificates/certificate_controller.go:43]: Describe what it does instead of what it holds

**Review by awly — CHANGES_REQUESTED** [review]

**Comment by zerodayz** [comment]
Thank you @awly  I am still trying to understand parts of kubernetes code, so please correct me if I am wrong.

```
// IgnorableError returns an error that we shouldn't handle (i.e. log) because
// it's spammy and usually user error. Instead we will log these errors at a
// higher log level. We still need to throw these errors to signal that the
// sync should be retried.
func IgnorableError(s string, args ...interface{}) ignorableError {
	return ignorableError(fmt.Sprintf(s, args...))
}
```

Which seems to be only called from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/certificates/approver/sarapprove.go#L112

And here we return an unexported variable and that's what golint doesn't like, but now my understanding is that, is exactly what we want from it. We only export the function so it can be called from other package but we can't call the ignorableError directly.

When you said it should return regular error instead, did you mean something like this:

```
func IgnorableError(s string, args ...interface{}) error {
	return fmt.Errorf(s, args...)
}
```

I tried changing it and it returned as Error type while when using ignorableError, it returned as ignorableError type:

`*errors.errorString`
VS
`.ignorableError`

return ignorableError(fmt.Sprintf(s, args...))

I am not so much familiar with the code base but wonder, if this could cause any problems if we change it to regular error.

Thank you for your time reviewing this and answer.

**Comment by awly** [comment]
@zerodayz it should look like so:

```
func IgnorableError(s string, args ...interface{}) error {
	return ignorableError(fmt.Sprintf(s, args...))
}
```

Which will work since `ignorableError` implements the `error` interface.
That way we both keep the internal error implementation hidden from the caller and keep the package API clean.

> One more question about the controller, do you know if there is any documentation to replace controller with custom controller, for debugging purposes?

This controller implementation is baked-in at compile time.
What are you trying to debug exactly? The answer will depend on that.

**Comment by zerodayz** [comment]
> @zerodayz it should look like so:
> 
> 
> 
> ```
> 
> func IgnorableError(s string, args ...interface{}) error {
> 
> 	return ignorableError(fmt.Sprintf(s, args...))
> 
> }
> 
> ```
> 
> 
> 
> Which will work since `ignorableError` implements the `error` interface.
> 
> That way we both keep the internal error implementation hidden from the caller and keep the package API clean.
> 
> 
> 
> > One more question about the controller, do you know if there is any documentation to replace controller with custom controller, for debugging purposes?
> 
> 
> 
> This controller implementation is baked-in at compile time.
> 
> What are you trying to debug exactly? The answer will depend on that.

I am trying to debug controller’s workflow, instead of reading thru the code, I want to run it. It will be easier to also understand what the original developer of the code meant. Some of the code just doesn’t give sense to me and I want to understand what exactly is going on, without having to ask everytime.

I am already using kind but was wondering if there is a better way just to compile cert controller and replace image in deployment?

I know that I can use kubebuilder or operator sdk for custom controller, but I am unsure about controllers that are already part of k8s.

Thank you for your help and your time with me. I will do the changes and run tests and see if there are any problems, will submit new updated commit.

**Comment by awly** [comment]
If I remember correctly, this controller runs in the [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) process.
One option is to re-compile and re-deploy `kube-controller-manager` (which sounds like what you've been doing so far).

The other option is to pass `--controllers=-csrapproving-csrsigning` flag to it (to disable the certificates controller) and run a separate controller process that you create. That separate process will take some work to create initially, but might be easier to iterate with.

File: pkg/controller/certificates/certificate_controller.go
Comment (zerodayz) [reply pkg/controller/certificates/certificate_controller.go:214]: Thank you for your review and comments, I have modified the ignorableError so it only returns error.

**Comment by zerodayz** [comment]
Thank you @awly for your help,

> If I remember correctly, this controller runs in the [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) process.
> One option is to re-compile and re-deploy `kube-controller-manager` (which sounds like what you've been doing so far).

I have been so far only re-compiling and re-deploying entire K8s cluster (using kind) not just individual component. I was thinking maybe there is an easier way, just like what you mentioned. (re-compile and re-deploy only kube-controller-manager) and replace in the deployment. Would that be possible?

> The other option is to pass `--controllers=-csrapproving-csrsigning` flag to it (to disable the certificates controller) and run a separate controller process that you create. That separate process will take some work to create initially, but might be easier to iterate with.

I am able to do the second part of running the controller separately from the cluster, but still have to find automated way to disable controllers (if possible with kind) or anything else. Want to avoid manual steps as much as possible.

**Comment by awly** [comment]
Sorry @zerodayz, I have not used kind and don't have any good recommendations on how do iterate more quickly.

**Comment by awly** [comment]
Ah, I see you've addressed the PR comments already.
Could you rebase the PR branch please?

**Comment by zerodayz** [comment]
> Ah, I see you've addressed the PR comments already.
> Could you rebase the PR branch please?

Finished and also fixed another introduced golint issue, i guess since package was in .golint_failures there was no linting and new commits in certificate controller package

**Comment by neolit123** [comment]
/remove-area apiserver cloudprovider code-generation conformance dependency e2e-test-framework ipvs kubeadm kubectl kubelet provider/gcp release-eng test
/remove-sig api-machinery apps architecture autoscaling cli cloud-provider cluster-lifecycle instrumentation network node release scheduling storage testing

**Pull request closed without merging**
